### PR TITLE
Clean the ngopt family

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

The NGOpt family is made of various improvements by plenty of distinct users -- let us clean a little bit.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/optimizerlib.py [hunk 1]
```diff
@@ -1959,7 +1960,7 @@ class NGOpt2(NGOptBase):
 
     def _select_optimizer_cls(self) -> base.OptCls:
         # override because it is different from NGOptBase (bug?)
-        self.has_discrete_not_softmax = self._has_discrete
+        #self.has_discrete_not_softmax = self._has_discrete
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1963]: if you want to remove it please just delete this line. Same with the other below
Just FYI this was a hack to stick with the original version

File: nevergrad/optimization/optimizerlib.py [hunk 2]
```diff
@@ -2045,10 +2046,12 @@ def _select_optimizer_cls(self) -> base.OptCls:
             else:
                 if self.has_noise and self.fully_continuous:
                     if budget > 100:
-                        optimClass = SQP
+                        if self.small_noise:
+                            optimClass = OnePlusOne
+                        else:
+                            optimClass = SQP
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:2052]:
```suggestion
                        optimClass = OnePlusOne if self.small_noise else SQP
```

File: nevergrad/optimization/optimizerlib.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py:1963]: Done.

Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py]: Done.

File: nevergrad/optimization/optimizerlib.py [hunk 3]
```diff
@@ -1875,6 +1875,7 @@ def __init__(self, parametrization: IntOrParameter, budget: tp.Optional[int] = N
         super().__init__(parametrization, budget=budget, num_workers=num_workers)
         descr = self.parametrization.descriptors
         self.has_noise = not (descr.deterministic and descr.deterministic_function)
+        self.small_noise = self.has_noise and descr.deterministic_function  # The noise coming from discrete variables goes to 0.
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1878]: any better name? small noise seems really weird and not helping for understanding the code. Maybe not naming it is actually more readable, why not just use has_noise & deterministic_function?

Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py]: I've switched to something explicit, but it's also a part of "small noise", i.e. noise which vanishes around the optimum.
An issue is that typically we rarely know that noise will vanish around the optimum --- except in the present case.

File: nevergrad/optimization/optimizerlib.py [hunk 4]
```diff
@@ -1920,7 +1921,7 @@ def _select_optimizer_cls(self) -> base.OptCls:
                                 cls = OnePlusOne if self.dimension > 30 else Cobyla
                             else:
                                 # DE is great in such a case (?).
-                                cls = DE if self.dimension > 2000 else CMA
+                                cls = DE if self.dimension > 2000 else CMA if self.dimension > 1 else OnePlusOne
```
Comment (teytaud) [inline nevergrad/optimization/optimizerlib.py:1924]: Just because CMA does weird things in dimension 2.

**Review by jrapin — APPROVED** [review]

**Pull request merged**